### PR TITLE
LibJS: Annotate Cell Allocation for ASAN

### PR DESCRIPTION
**Pull request by ADKaster**
Annotate heap block allocations for address sanitizer.

Mark the entirety of a heap block's storage poisoned at construction.
Unpoison all of a Cell's memory before allocating it, and re-poison as
much as possible on deallocation. Unfortunately, the entirety of the
FreelistEntry must be kept unpoisoned in order for reallocation to work
correctly.

Decreasing the size of FreelistEntry or adding a larger redzone to Cells
would make the instrumentation even better.

### CODE REVIEW

File: Userland/Libraries/LibJS/Heap/HeapBlock.h [hunk 1]
```diff
@@ -27,13 +32,17 @@ class HeapBlock {
 
     ALWAYS_INLINE Cell* allocate()
     {
+        Cell* allocated_cell = nullptr;
         if (m_freelist) {
             VERIFY(is_valid_cell_pointer(m_freelist));
-            return exchange(m_freelist, m_freelist->next);
+            allocated_cell = exchange(m_freelist, m_freelist->next);
+        } else if (has_lazy_freelist())
+            allocated_cell = cell(m_next_lazy_freelist_index++);
```
Comment (linusg) [inline Userland/Libraries/LibJS/Heap/HeapBlock.h:40]: Please don't omit curly braces if any previous or following if/else branch needs them.

File: Userland/Libraries/LibJS/Heap/HeapBlock.h [hunk 2]
```diff
@@ -96,16 +96,14 @@ class HeapBlock {
         return reinterpret_cast<Cell*>(&m_storage[index * cell_size()]);
     }
 
-    FreelistEntry* init_freelist_entry(size_t index)
-    {
-        return new (&m_storage[index * cell_size()]) FreelistEntry();
-    }
```
Comment (awesomekling) [inline Userland/Libraries/LibJS/Heap/HeapBlock.h:102]: This could just as well be a separate commit :^)

File: Userland/Libraries/LibJS/Heap/HeapBlock.h (hunk 1 above)
Comment (ADKaster) [reply Userland/Libraries/LibJS/Heap/HeapBlock.h]: Ah fair. I wonder if there's a way to get the clang-format config to enforce this rule?

File: Userland/Libraries/LibJS/Heap/HeapBlock.h (hunk 2 above)
Comment (ADKaster) [reply Userland/Libraries/LibJS/Heap/HeapBlock.h:102]: Good point, drive-by is not a good thing. I split it out.
